### DISCUSSION TITLE
'Grabbing' a z-wave message when sent from the controller

### DISCUSSION THREAD

**Discussion by lurendrejer** [zwave-js/node-red-contrib-zwave-js#139] [Q&A] [answered]
I'm tinkering with indicator messages - I'm trying to create a 'base' message in a function.

When I connect a debug node directly to the controller node and send an 'indicator multilevel 50' from the gui, I actually expected to see the message in the debug panel - but didn't.
Can I capture messages in this manner? To create a first, basic test message?
I've been fiddling with the 'expand' CC-flag, etc. but I'm not really getting anywhere.

ps. the device updates it's indicator just fine - I would just like to know which command was sent to the device and which parameters was included.

**Comment by marcus-j-davies** [comment, marked as answer]
Hey,

Are you referring to the **Indicator** CC or  **Multilevel Switch** CC - if so The Indicator  CC does not 'respond' like other CCs'
(therefore we don't see anything) its really only to... well.... indicate - i.e some devices use them to set an LED,  some devices use them to trigger a sound, or in the case of outdoor sirens - to illuminate the logo on the box.

Do you have a manual for the device?

**Reply by marcus-j-davies** [reply]
I use Indicator (well, my device does) to highlight the current security mode on my alarm (its a security Pin pad)

**Reply by lurendrejer** [reply]
Hi.
It is the indicator CC.
![image](https://user-images.githubusercontent.com/30922656/137640423-212c4b11-25a6-4083-b8e1-5b19e3f19088.png)

I just figured that since the controller sent the command, it would show up in the debug node :)
I don't have much luck using the cmd-factory node and wanted to see what a working message actually looked like.

Next problem will be how to send a "blink" command, since the device supports that - it might just be to send 255 or something, I'll figure that out later.

**Reply by lurendrejer** [reply]
That is more or less what i'll be using it for.
This particular indicator sets the status of the garage - open/closed.

**Reply by marcus-j-davies** [reply]
All output is only for messages we receive (not what we send).

HOWEVER! there is a way to see the command.

1) Open up your browser debug panel
2) Monitor network commands (clear them down first)
3) send command via the UI
4) you will see it in network monitor.

Indicator has gone through some revisions with Zwave Alliance - V1 was very simple, V2+ introduces more advanced control.

**Reply by marcus-j-davies** [reply]
ignore the **noWait** stuff - that is specific to the UI

**Reply by marcus-j-davies** [reply]
Example:
This is what I see, when I set an indicator value in the UI.
![Screenshot 2021-10-17 at 20 01 02](https://user-images.githubusercontent.com/55892693/137641206-2b908185-a681-4558-8fea-dec06a1dea9e.png)

**Reply by lurendrejer** [reply]
Ok, thank you - I'm held up right now, I'll get back as soon as possible - thank you, again! :)

**Reply by marcus-j-davies** [reply]
Sure thing.

its likely using V2+ on the Indicator - which is easy enough on the factory.
The UI is using the ValueAPI - which again you can use on the factory also..

or just go the route of constructing the message yourself - just remember to take out **noWait** if 'steeling' it from network log.

**Comment by lurendrejer** [comment]
I should be able to test this today/night - I'm very appreciative of your help, just hung up like hell...

I had great fun with indicators for a few hours yesterday - and will report back as soon as I have the time.

**Comment by marcus-j-davies** [comment]
Yeah, the Indicator CC is a good one,
a lot of outdoor sirens use it - to illuminate the logo, or to strobe during an event.

My security pin pad uses it to indicate the current security arm modes (or to create a countdown i.e to exit the property for an example) - so I can easily sync it with my DIY alarm setup.

Again, everything can be done within factory, just call out if you need help.

No worries. - I am usually around.

**Comment by lurendrejer** [comment]
It worked i'm no longer stranded.
Thank you for the 'push' :)

{mode: "ValueAPI", method: "setValue", node: 4,…}
method: "setValue"
mode: "ValueAPI"
noWait: true
node: 4
params: [{commandClassName: "Indicator", commandClass: 135, endpoint: 0, property: 69, propertyKey: 1,…}, 1]
0: {commandClassName: "Indicator", commandClass: 135, endpoint: 0, property: 69, propertyKey: 1,…}
1: 1

**Comment by marcus-j-davies** [comment]
before sure to remove **noWait: true**
that is strictly for the UI

**Comment by lurendrejer** [comment]
This was the final working result - a multilevel+indicator sender in an easy understandable subflow :)

I'll be tinkering with the blinking same time tomorrow (the device has blinking and blinking speed afaik)

`[
    {
        "id": "a94ded8d03c5e1d9",
        "type": "subflow",
        "name": "Z-Wave Multilevel",
        "info": "Craft Binary and Basic Set - commands",
        "category": "ZWave JS",
        "in": [
            {
                "x": 40,
                "y": 160,
                "wires": [
                    {
                        "id": "e5079836e6e1d271"
                    },
                    {
                        "id": "67270eb4c3f814ab"
                    }
                ]
            }
        ],
        "out": [
            {
                "x": 1160,
                "y": 120,
                "wires": [
                    {
                        "id": "058cd0bbbd09bbd9",
                        "port": 0
                    },
                    {
                        "id": "d6cc163e3ed44aa6",
                        "port": 0
                    }
                ]
            }
        ],
        "env": [
            {
                "name": "cc",
                "type": "str",
                "value": "Indicator",
                "ui": {
                    "icon": "font-awesome/fa-creative-commons",
                    "label": {
                        "en-US": "Command"
                    },
                    "type": "select",
                    "opts": {
                        "opts": [
                            {
                                "l": {
                                    "en-US": "Multilevel Switch"
                                },
                                "v": "Multilevel Switch"
                            },
                            {
                                "l": {
                                    "en-US": "Indicator"
                                },
                                "v": "Indicator"
                            }
                        ]
                    }
                }
            },
            {
                "name": "endpoint",
                "type": "num",
                "value": "",
                "ui": {
                    "icon": "font-awesome/fa-cogs",
                    "type": "input",
                    "opts": {
                        "types": [
                            "num"
                        ]
                    }
                }
            },
            {
                "name": "state",
                "type": "num",
                "value": "0",
                "ui": {
                    "icon": "font-awesome/fa-lightbulb-o",
                    "label": {
                        "en-US": "State"
                    },
                    "type": "spinner",
                    "opts": {
                        "min": 0,
                        "max": 99
                    }
                }
            }
        ],
        "meta": {},
        "color": "#C0DEED",
        "inputLabels": [
            "Any trigger"
        ],
        "outputLabels": [
            "Z-wave binary switch"
        ],
        "icon": "font-awesome/fa-send",
        "status": {
            "x": 460,
            "y": 200,
            "wires": [
                {
                    "id": "67270eb4c3f814ab",
                    "port": 0
                }
            ]
        }
    },
    {
        "id": "058cd0bbbd09bbd9",
        "type": "change",
        "z": "a94ded8d03c5e1d9",
        "name": "Endpoint",
        "rules": [
            {
                "t": "set",
                "p": "payload.endpoint",
                "pt": "msg",
                "to": "endpoint",
                "tot": "env"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 980,
        "y": 140,
        "wires": [
            []
        ]
    },
    {
        "id": "e5079836e6e1d271",
        "type": "function",
        "z": "a94ded8d03c5e1d9",
        "name": "TemplateCC",
        "func": "let Message = {\n    payload: {\n        mode: \"CCAPI\",\n        method: \"set\",\n    }\n}\nreturn Message",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 170,
        "y": 120,
        "wires": [
            [
                "c0ad86a69d1569a3"
            ]
        ]
    },
    {
        "id": "c0ad86a69d1569a3",
        "type": "change",
        "z": "a94ded8d03c5e1d9",
        "name": "Type",
        "rules": [
            {
                "t": "set",
                "p": "payload.cc",
                "pt": "msg",
                "to": "cc",
                "tot": "env"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 350,
        "y": 120,
        "wires": [
            [
                "06bd7ddd16b15907"
            ]
        ]
    },
    {
        "id": "d6cc163e3ed44aa6",
        "type": "change",
        "z": "a94ded8d03c5e1d9",
        "name": "NoEndpoint",
        "rules": [
            {
                "t": "delete",
                "p": "payload.endpoint",
                "pt": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 990,
        "y": 100,
        "wires": [
            []
        ]
    },
    {
        "id": "14a760e4fe80b96a",
        "type": "switch",
        "z": "a94ded8d03c5e1d9",
        "name": "",
        "property": "endpoint",
        "propertyType": "env",
        "rules": [
            {
                "t": "eq",
                "v": "0",
                "vt": "num"
            },
            {
                "t": "else"
            }
        ],
        "checkall": "false",
        "repair": false,
        "outputs": 2,
        "x": 730,
        "y": 120,
        "wires": [
            [
                "d6cc163e3ed44aa6"
            ],
            [
                "058cd0bbbd09bbd9"
            ]
        ]
    },
    {
        "id": "67270eb4c3f814ab",
        "type": "change",
        "z": "a94ded8d03c5e1d9",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "{}",
                "tot": "json"
            },
            {
                "t": "set",
                "p": "payload.state",
                "pt": "msg",
                "to": "state",
                "tot": "env"
            },
            {
                "t": "set",
                "p": "payload.endpoint",
                "pt": "msg",
                "to": "endpoint",
                "tot": "env"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 260,
        "y": 200,
        "wires": [
            []
        ]
    },
    {
        "id": "06bd7ddd16b15907",
        "type": "change",
        "z": "a94ded8d03c5e1d9",
        "name": "State",
        "rules": [
            {
                "t": "set",
                "p": "payload.params[0]",
                "pt": "msg",
                "to": "state",
                "tot": "env"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 530,
        "y": 120,
        "wires": [
            [
                "14a760e4fe80b96a"
            ]
        ]
    }
]`

**Comment by marcus-j-davies** [comment]
very nice,
picking up (or removing parts) of the message along the way - Nice!

**Comment by lurendrejer** [comment]
Well, the cmd-factory node is way too complicated for me.
I'm hoping that my step-by-step approach will end up in me understanding the more advaned cmd-factory node down the line.

The sheer amount of inputs in the cmd-factory makes the learning curve very steep.

I have such flows for basic/binary set, scenes and now indicator/multichannel - they get the job done.

And thank you, very much - I'm pretty proud of the progress i'm making 😃

**Comment by lurendrejer** [comment]
hmm
![image](https://user-images.githubusercontent.com/30922656/137793954-ce6b0c71-ca26-49af-99c3-6270beedc6aa.png)

**Comment by marcus-j-davies** [comment]
So with Indicator V2+
You have the the **indicator ID** - this is the actual indicator that is to be affected.
Then you have a **Property ID**:

0x01: The Multi Level Value
0x02: The Binary Value
0x03: The ON/Off Period
0x04: The ON/Off Cycles

Then the **value** which can be numeric for **0x01**, **0x03** and **0x04**
or `true`|`false` for **0x02**

**Comment by marcus-j-davies** [comment]
So,

to set 2 properties,

```javascript
let  Multilevel =  {
	indicatorId: 1,
	propertyId: 1,
	value: 30 // Brightness I assume
}

let  OnOffCycles =  {
	indicatorId: 1,
	propertyId: 4,
	value: 5 // How many times it flickers?
}

let Props = [Multilevel, OnOffCycles]

let Message = {
    payload: {
        mode: "CCAPI",
        node: 2,
        cc: "Indicator",
        method: "set",
         /* This needs to be an array, but setting using 2 indicator properties, in its self needs to be an array ;-) */
         /* so param 0 is actually an array ;-) */
         params: [Props]
    }
}
return Message
```


https://zwave-js.github.io/node-zwave-js/#/api/CCs/Indicator?id=set

**Comment by lurendrejer** [comment]
I'm trying to mark this as answered - but my morning-IQ is not sufficient.

//Edit: I wanted to mark the Marcus' post about browser/network monitoring as the answer, but couldn't. I marked the post above it instead.

**Comment by marcus-j-davies** [comment]
@lurendrejer 

Someone once asked me, if it were possible to view the messages being sent from the UI - so you can capture the required message.

Well....Someone once implemented the feature.

@crxporter  - Thoughts as well?

NOTE: This is only for UI interaction - its not to monitor everything being sent out to the network, it only details what commands from the UI are taking place - but you can use the message being recorded in your setup later.

https://user-images.githubusercontent.com/55892693/138511095-1bfb2c6e-6ca3-4baa-9d43-abbc2f1d422a.mov

**Reply by marcus-j-davies** [reply]
This is just a "is it useful" type thing.
So I'll await both your feedback.

**Reply by thk-socal** [reply]
There are some places where that would be very awesome to see. I think "it is useful" and would keep it.

**Comment by lurendrejer** [comment]
The ability to capture both value and cc API would be very helpful in my usecase.

I actually thought I could use the GUI to build commands like that already (expand CC -option).

The command viewer looks great - I would welcome it, greatly.

**Reply by marcus-j-davies** [reply]
The UI is the only place I can capture the comms - but the UI uses the ValueID, so it will be limited to ValueAPI.
My plan is, to make it more 'log' like - so you have a list of the commands that has taken place.

Expand CCs simply auto expands the CC list when selecting a node - nothing special here sadley 😞

**Comment by marcus-j-davies** [comment]
Ok,

Having it log like - seems a really nice way?

![image](https://user-images.githubusercontent.com/55892693/138517552-2205b6e0-245b-441d-8657-47c5fdf281c5.png)

**Reply by lurendrejer** [reply]
Absolutely

**Reply by marcus-j-davies** [reply]
Done! - it auto scrolls

**Reply by marcus-j-davies** [reply]
Its also placed in the right location.

![image](https://user-images.githubusercontent.com/55892693/138521312-dad94e6d-d2e2-421c-8cba-03bd93400696.png)

**Reply by crxporter** [reply]
> Its also placed in the right location.

Could be "include/exclude" in one row with "UI monitor" in the same button row? We don't need the vertical growth again

**Reply by lurendrejer** [reply]
I would probably have gone with a flag on the controller (using the debug panel) or something but I don't have the same context as you. 
It probably makes sense to have it in the menu.
The thing I was looking for yesterday was more a less a copy to clipboard with the target value. But I know that is is a very specific use case.

The log-type display and button in the menu is great for me and other new people.
As programmers you might be blind to how much of a hard time people are having with JSON, arrays, etc. :)

I'm at a point where I probably don't even know, that I don't know JSON - it works and I can debug it, but I don't really understand it.

**Reply by marcus-j-davies** [reply]
> Could be "include/exclude" in one row with "UI monitor" in the same button row? We don't need the vertical growth again

Fully agree here

> I would probably have gone with a flag on the controller (using the debug panel) or something but I don't have the same context as you.

This is my thought: A lot of users ~~wil first be finding~~ find out very quickly they can control their devices via the UI, and will immediately, want to know what the command is - so they can automate the value changes they have done in the UI.

This is exactly what the new "UI Monitor" can do, the commands you see are directly usable in function nodes, where you can change any aspect of the command once copied - i.e the value to actually use for example.

I have omitted inclusion/exclusion commands - as they are not simple commands, and require various steps to do a proper inclusion.

![image](https://user-images.githubusercontent.com/55892693/138545494-9a92a839-6161-45ac-9baf-1580f390f173.png)

Before S2 was introduced, it was possible to do it with commands - Hey! We did once need to 😁 

I did think about using the debug panel - but we cant see it if in the UI, the controller in your flow is not connected directly to the UI, which makes it more difficult to log the commands on the controller node its self.

**Reply by lurendrejer** [reply]
Makes perfect sense :)

**Comment by marcus-j-davies** [comment]
The finished UI Monitor.
I have styled it up a little - which really completes it I think.

https://user-images.githubusercontent.com/55892693/138546597-64a4b260-499b-47d8-b878-077efba0b50e.mov

**Reply by lurendrejer** [reply]
Nice!

**Reply by marcus-j-davies** [reply]
**6.3.0** has now been released - it should appear in the palette menu within an hour.
shout if any issues (if/when you update) 

Change Log:
https://github.com/zwave-js/node-red-contrib-zwave-js/blob/main/CHANGELOG.md

**Reply by lurendrejer** [reply]
just upgraded - no fires yet.

**Reply by marcus-j-davies** [reply]
Give it time! 😉

**Reply by marcus-j-davies** [reply]
Hi @lurendrejer 

Every so often (after lengthy discussions), I like to find out if all is well?
assuming 'no fires yet.' still stands true?

if all is well, just reach out to us, when assistance is required.

**Reply by lurendrejer** [reply]
Everything is dandy - no fires what so ever!
I'm pushing everyone in my social circle towards zwave-js 😅

**Reply by marcus-j-davies** [reply]
We have some more on the way, to support a Security S2 feature.

Scanning a QR code to include a device.
Some devices have what is called **Smart Start** - a kind of Scan and forget type deal.

Scan the devices QR code, and forget about it, these devices will advertise them self, and scanning its QR code, the module will 'Listen' for this advertisement, once the device is turned on / powered up - it will auto include.

Its ready to use from the driver - but still has some kinks that are being ironed out.
once all stable - I will include this feature.

I will be implementing a web ui, allowing you to walk around the house scanning your devices to include with your phone.

**Reply by lurendrejer** [reply]
This is getting ridiculous - you guys have better performance and compatibility than any of the old players in the game, already. X-D

**Reply by thk-socal** [reply]
@lurendrejer this is why I moved all my zwave over. Performance is amazing and I can now do so much more with automation and node-red and even inside of Homekit with what @marcus-j-davies has created. Welcome aboard and enjoy the ride!

**Reply by marcus-j-davies** [reply]
............ That's better, sorry, my head needed shrinking back down 😉 

Thanks guys, but I cant take all the credit,  @AlCalzone, who develops/maintains the backend z-wave javascript driver that this module uses, is the reason its possible.

There are a few of us who are under the zwave-js umbrella, we all maintain the different parts, but the main driver is AI's work.
Of course not forgetting all the fabulous users of this framework, who shape it all 👍
